trie-db/src/node.rs

```diff
@@ -40,21 +41,52 @@ pub fn decode_hash<H: Hasher>(data: &[u8]) -> Option<H::Out> {
 	Some(hash)
 }
 
+/// Value representation in `Node`.
+#[derive(Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Value<'a> {
+	/// Node with no value attached.
+	NoValue,
+	/// Value byte slice.
+	Value(&'a [u8]),
+	/// Hash byte slice and the actual value if accessed.
+	HashedValue(&'a [u8], Option<DBValue>),
+}
+
+impl<'a> Value<'a> {
+	pub(crate) fn new(value: Option<&'a [u8]>, threshold: Option<u32>) -> Option<Self> {
+		Some(match value {
+			Some(value) => {
+				if let Some(threshold) = threshold {
+					if value.len() >= threshold as usize {
+						return None;
+					} else {
+						Value::Value(value)
+					}
+				} else {
+					Value::Value(value)
+				}
+			},
+			None => Value::NoValue,
+		})
+	}
+}
+
 /// Type of node in the trie and essential information thereof.
 #[derive(Eq, PartialEq, Clone)]
 #[cfg_attr(feature = "std", derive(Debug))]
 pub enum Node<'a> {
 	/// Null trie node; could be an empty root or an empty branch entry.
 	Empty,
 	/// Leaf node; has key slice and value. Value may not be empty.
-	Leaf(NibbleSlice<'a>, &'a [u8]),
+	Leaf(NibbleSlice<'a>, Value<'a>),
```

Comment:
in the past the leaf node couldn't contain no value. Now it can, also can a leaf node contain a hashed value ?
It seems to me this should use another type LeafValue with only 1 or 2 variant (depending if it can contains hashed value which I don't know yet).

Reply:
Leaf can contain hashed value (it is as needed as for the branch case).
But by definition a Leaf cannot contain no value (otherwhise it would be removed).

Most of the code for Leaf without value leads to panic (unreachable probably, I did not really check), or at least it cannot encode or decode.
I did not want to have two different Value enum (we got other enums for Value when data is owned or for decode plan).
But now that the code is a bit more reviewed, maybe using `LeafValue` could make sense, would just make more code.

Reply:
we could also keep the old way and have Value here, and `Option<Value>` for branches maybe. and remove the variant NoValue

Reply:
I find it better now